src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -1207,6 +1217,53 @@ private Task WriteStringAsync(string s, bool async)
             return WriteStringAsyncSlow(s, async);
         }
 
+        private Task WriteStringAsync(string s, Encoding? encoding, bool async)
+        {
+            if (encoding is null)
+            {
+                return WriteStringAsync(s, async);
+            }
+
+            // If there's enough space in the buffer to just copy all of the string's bytes, do so.
+            int offset = _writeOffset;
+            int available = _writeBuffer.Length - offset;
+
+            // Fast-path Latin1 and UTF8 as the most common scenarios of non-default encoding
+            const int MaxUtf8BytesPerChar = 3;
+            bool sufficientBufferSpace = ReferenceEquals(encoding, Encoding.Latin1) ? available >= s.Length :
+                ReferenceEquals(encoding, Encoding.UTF8) ? available >= s.Length * MaxUtf8BytesPerChar :
+                available >= encoding.GetMaxByteCount(s.Length);
+
+            if (sufficientBufferSpace)
```

Comment:
Smaller code gen with this:

```suggestion
            int requiredSpace = ReferenceEquals(encoding, Encoding.Latin1) ? s.Length :
                ReferenceEquals(encoding, Encoding.UTF8) ? s.Length * MaxUtf8BytesPerChar :
                encoding.GetMaxByteCount(s.Length);

            if (available >= requiredSpace)
```

Reply:
Changed